app/gui2/src/stores/graph/graphDatabase.ts

```diff
@@ -1,54 +1,142 @@
 import { SuggestionDb, groupColorStyle, type Group } from '@/stores/suggestionDatabase'
 import { tryGetIndex } from '@/util/array'
 import { Ast, AstExtended } from '@/util/ast'
+import { AliasAnalyzer } from '@/util/ast/aliasAnalysis'
 import { colorFromString } from '@/util/colors'
 import { ComputedValueRegistry, type ExpressionInfo } from '@/util/computedValueRegistry'
+import { MappedKeyMap } from '@/util/containers'
 import { ReactiveDb, ReactiveIndex, ReactiveMapping } from '@/util/database/reactiveDb'
 import type { Opt } from '@/util/opt'
 import { qnJoin, tryQualifiedName } from '@/util/qualifiedName'
 import { Vec2 } from '@/util/vec2'
+import { iteratorFilter } from 'lib0/iterator'
 import * as set from 'lib0/set'
 import {
   IdMap,
   visMetadataEquals,
+  type ContentRange,
   type ExprId,
   type NodeMetadata,
   type VisualizationMetadata,
 } from 'shared/yjsModel'
 import { ref, type Ref } from 'vue'
 
-export class GraphDb {
-  nodes = new ReactiveDb<ExprId, Node>()
-  idents = new ReactiveIndex(this.nodes, (_id, entry) => {
-    const idents: [ExprId, string][] = []
-    entry.rootSpan.visitRecursive((span) => {
-      if (span.isTree(Ast.Tree.Type.Ident)) {
-        idents.push([span.astId, span.repr()])
-        return false
+export interface AliasInfo {
+  identifier: AstExtended
+  usages: Set<ExprId>
+}
+
+export class ConnectionsDb {
+  aliases = new ReactiveDb<ExprId, AliasInfo>()
+  identifiers = new ReactiveIndex(this.aliases, (alias, info) => [[info.identifier.repr(), alias]])
+
+  readFunctionAst(ast: AstExtended<Ast.Tree.Function>) {
+    const analyzer = new AliasAnalyzer(ast.parsedCode, ast.inner)
+    analyzer.process()
+    const aliasRangeToTree = new MappedKeyMap<ContentRange, AstExtended | undefined>(
+      IdMap.keyForRange,
+    )
+    const aliasIdToRange = new Map<ExprId, ContentRange>()
+    for (const [alias, usages] of analyzer.aliases) {
+      aliasRangeToTree.set(alias, undefined)
+      for (const usage of usages) aliasRangeToTree.set(usage, undefined)
+    }
+    ast.visitRecursive((ast) => {
+      if (aliasRangeToTree.has(ast.span())) {
+        aliasRangeToTree.set(ast.span(), ast)
+        aliasIdToRange.set(ast.astId, ast.span())
       }
       return true
     })
-    return idents
+
+    for (const key of this.aliases.keys()) {
+      const range = aliasIdToRange.get(key)
+      if (range == null || !analyzer.aliases.has(range)) {
+        this.aliases.delete(key)
+      }
+    }
+
+    for (const [alias, usages] of analyzer.aliases) {
```

Comment:
For clarity, rename `alias` to ``binding`, or find another good not confusing name. Add TODO to make alias-analyzer names more clear.